rclcpp/src/rclcpp/signal_handler.cpp

```diff
@@ -242,143 +255,79 @@ SignalHandler::get_old_signal_handler(int signum)
-    1,  // maximum semaphore count
-    NULL);  // unnamed semaphore
-  if (NULL == signal_handler_sem_) {
-    throw std::runtime_error("CreateSemaphore() failed in setup_wait_for_signal()");
-  }
-#elif defined(__APPLE__)
-  signal_handler_sem_ = dispatch_semaphore_create(0);
-#else  // posix
-  if (-1 == sem_init(&signal_handler_sem_, 0, 0)) {
-    throw std::runtime_error(std::string("sem_init() failed: ") + strerror(errno));
-  }
-#endif
-  wait_for_signal_is_setup_.store(true);
-}
-
-void
-SignalHandler::teardown_wait_for_signal() noexcept
-{
-  if (!wait_for_signal_is_setup_.exchange(false)) {
-    return;
-  }
-#if defined(_WIN32)
-  CloseHandle(signal_handler_sem_);
-#elif defined(__APPLE__)
-  dispatch_release(signal_handler_sem_);
-#else  // posix
-  if (-1 == sem_destroy(&signal_handler_sem_)) {
-    RCLCPP_ERROR(get_logger(), "invalid semaphore in teardown_wait_for_signal()");
-  }
-#endif
-}
-
-void
-SignalHandler::wait_for_signal()
-{
-  if (!wait_for_signal_is_setup_.load()) {
-    RCLCPP_ERROR(get_logger(), "called wait_for_signal() before setup_wait_for_signal()");
-    return;
-  }
-#if defined(_WIN32)
-  DWORD dw_wait_result = WaitForSingleObject(signal_handler_sem_, INFINITE);
-  switch (dw_wait_result) {
-    case WAIT_ABANDONED:
-      RCLCPP_ERROR(
-        get_logger(), "WaitForSingleObject() failed in wait_for_signal() with WAIT_ABANDONED: %s",
-        GetLastError());
-      break;
-    case WAIT_OBJECT_0:
-      // successful
+  switch(input) {
+    case Input::SigInt:
+      got_sig_int.exchange(true);
       break;
-    case WAIT_TIMEOUT:
-      RCLCPP_ERROR(get_logger(), "WaitForSingleObject() timedout out in wait_for_signal()");
+    case Input::SigTerm:
+      got_sig_term.exchange(true);
       break;
-    case WAIT_FAILED:
-      RCLCPP_ERROR(
-        get_logger(), "WaitForSingleObject() failed in wait_for_signal(): %s", GetLastError());
+    case Input::TerminateHandler:
+      terminate_handler_.exchange(true);
       break;
-    default:
-      RCLCPP_ERROR(
-        get_logger(), "WaitForSingleObject() gave unknown return in wait_for_signal(): %s",
-        GetLastError());
   }
-#elif defined(__APPLE__)
-  dispatch_semaphore_wait(signal_handler_sem_, DISPATCH_TIME_FOREVER);
-#else  // posix
-  int s;
-  do {
-    s = sem_wait(&signal_handler_sem_);
-  } while (-1 == s && EINTR == errno);
-#endif
-}
 
-void
-SignalHandler::notify_signal_handler() noexcept
-{
-  if (!wait_for_signal_is_setup_.load()) {
-    return;
-  }
-#if defined(_WIN32)
-  if (!ReleaseSemaphore(signal_handler_sem_, 1, NULL)) {
-    RCLCPP_ERROR(
-      get_logger(), "ReleaseSemaphore() failed in notify_signal_handler(): %s", GetLastError());
-  }
-#elif defined(__APPLE__)
-  dispatch_semaphore_signal(signal_handler_sem_);
-#else  // posix
-  if (-1 == sem_post(&signal_handler_sem_)) {
-    RCLCPP_ERROR(get_logger(), "sem_post failed in notify_signal_handler()");
-  }
-#endif
+  signal_conditional_.notify_one();
```

Comment:
Are you sure this is safe to do from a signal handler? I believe the custom code was used explicitly because there was no standard way to do this safely.

Reply:
Also, see the blame:

- https://github.com/ros2/rclcpp/pull/607
- https://github.com/ros2/rclcpp/pull/605#discussion_r240439968